### DISCUSSION TITLE
Not Null constraint for migration to 2.0+

### DISCUSSION THREAD

**Discussion by FranckRodriguez** [sqlalchemy/sqlalchemy#11880] [Usage Questions]
Hello,

I'm trying to migrate from1.4 to last 2.0+.

Following the migration guide for my mapper, I'm not sure how nullable parameter works.

My new mapper looks like :
```py
class My_Table(Base):
    __tablename__ = "My_Table"

    # Classic column
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String(50))
    other_table_id: Mapped[int] = mapped_column(Integer, ForeignKey("Other_Table.id"))

    __table_args__ = {'sqlite_autoincrement': True}

    # Relation ManyToOne
    Other_Table = relationship('Other_Table',
                               backref=backref('My_Table'),
                               lazy="joined")
```

Using this, all my table classic parameters (not considering foreign keys) gives me NOT NULL constraint when empty. 
I can't find a way to set nullable parameter to dafault=True, to avoid redefining all my Mapped[str] to Mapped[Optional[str]].

All (almost) my Mapped[str] must be nullable...

**Comment by zzzeek** [comment]
if they're nullable then you'd normally use `Optional[str]` or more succinctly `str | None` in your types.   When you transitioned from `Column` to `mapped_column()` and added these new `Mapped` types, that's where you would have done that.

if you want all your Python typing to consider all these columns as never None but they are actually nullable in the database, are you sure they never come back as NULL in your code?   ~~you can just make a wrapper function like `nullable_mapped_column()` and do a search and replace on that function name~~  use an `Annotated` type with `mapped_column(nullable=False)` if you want this

**Reply by CaselIT** [reply]
You could also customize the type using `Mapped[my_str]` where `my_str = Annotated[str, mapped_column(nullable=False)]`

**Reply by zzzeek** [reply]
oh right!!   :) duh who had that great idea!   (wasn't actually me :) )

**Reply by CaselIT** [reply]
don't remember actually. looking a bit at the changelog I can't pinpoint what issue added it

**Reply by FranckRodriguez** [reply]
I've added Mapped types to my mapper to avoid pycharm warnings in all my queries.

Using mapped_columns instead of Columns seems to set nullable to default False, and all my python test on my queries are KO (cause only requested info is set in the tested database).

When i was using Columns(String(50)), there was no NOT NULL constraints errors and I had to use nullable=False if wanted.
Same thing for some relationships Foreign keys which are not mandatory 

So, I understand I can use Mapped[Optional[...]] on my types, but can't i juste reverse the nullable default value to True like it was in 1.4?


(I apologize for my english, I try to be understandable)

**Reply by CaselIT** [reply]
> Using mapped_columns instead of Columns seems to set nullable to default False

The default in v2 is that if nullable is not provided, it tries to infer it from the type definition. `Mapped[str]` means that `None` is not permitted, so the column is set as not-null. If you use `Mapped[str|None]` instead None is allowed so the column will be set as nullable.

> but can't i just reverse the nullable default value to True like it was in 1.4?

Use something like `Mapped[my_str]` as mentioned above. But I would not advise is since you are basically adding wrong typing support.

Note that you don't need to use `Mapped` or `mapped_column` if you don't want to. It's just very useful if you plan on using type checkers etc compared to the old way of mapping the orm classes

**Reply by FranckRodriguez** [reply]
hmm okay. I wanted to use latest version to simplify maintenance on packages. But i guess i don't really need to do it.

Columns and v1.4 works fine with no types warnings x).

Will try Mapped[str | None] or Optional[str] later and close this issue if works well.

Thx!

**Reply by CaselIT** [reply]
you can use columns also in 2.0. It's still compatible with that format

**Reply by CaselIT** [reply]
By columns I mean that this will work as in 1.4 also in v2:
```py
class My_Table(Base):
    __tablename__ = "My_Table"

    # Classic column
    id = Column(Integer, primary_key=True)
    name = Column(String(50))
    other_table_id = Column(Integer, ForeignKey("Other_Table.id"))

    __table_args__ = {'sqlite_autoincrement': True}

    # Relation ManyToOne
    Other_Table = relationship('Other_Table',
                               backref=backref('My_Table'),
                               lazy="joined")
```

You will loose most type checking with this, but if you are fine with it feel free to keep using it.
There are currently no plan of deprecating this flavor of orm mapping
